core/sequencer/mapper/fns.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mapper contains a transformation pipeline from log messages to map revisions.
+package mapper
+
+import (
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+
+	"github.com/google/keytransparency/core/mutator"
+	"github.com/google/keytransparency/core/mutator/entry"
+
+	pb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
+	tpb "github.com/google/trillian"
+)
+
+// IndexedUpdate is a KV<Index, Update> type.
+type IndexedUpdate struct {
+	Index  []byte
+	Update *pb.EntryUpdate
+}
+
+// MapUpdateFn converts an update into an IndexedUpdate.
+func MapUpdateFn(msg *pb.EntryUpdate) (*IndexedUpdate, error) {
+	var e pb.Entry
+	if err := proto.Unmarshal(msg.Mutation.Entry, &e); err != nil {
+		return nil, err
+	}
+	return &IndexedUpdate{
+		Index:  e.Index,
+		Update: msg,
+	}, nil
+}
+
+// ReduceFn decides which of multiple updates can be applied in this revision.
+// TODO(gbelvin): Move to mutator interface.
+func ReduceFn(mutatorFn mutator.ReduceMutationFn,
+	index []byte, leaves []*tpb.MapLeaf, msgs []*pb.EntryUpdate, emit func(*tpb.MapLeaf)) {
+	if got := len(leaves); got != 1 {
+		glog.Errorf("expected 1 map leaf for index %x, got %v", index, got)
+		return
+	}
+	var oldValue *pb.SignedEntry // If no map leaf was found, oldValue will be nil.
+	var err error
+	if len(leaves) > 0 {
+		oldValue, err = entry.FromLeafValue(leaves[0].GetLeafValue())
+		if err != nil {
+			glog.Warningf("entry.FromLeafValue(): %v", err)
+			return
+		}
+	}
+
+	if len(msgs) == 0 {
+		return
```

Comment:
Can `msgs` be empty in normal operation, or it's an error/warning? Presumably the pipeline fetches only the indexes that are being updated, i.e. those that have non-empty `msgs`.

Reply:
`len(msgs)` means that the join step was provided a map value, but not a log item. 
In general this is permitted, but in our case it would be a surprise, so yes, we should error.